hnn_core/network.py

```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
             plt.show()
         return ax.get_figure()
 
-    def plot_spikes(self, ax=None, show=True):
-        """Plot the spiking activity for each cell type.
+
+class Spikes(object):
+    '''The Spikes class.
+
+    Parameters
+    ----------
+    times : tuple (n_trials, ) of list of float | None
+        Each element of the tuple is a trial.
+        The list contains the time stamps of spikes.
+    gids : tuple (n_trials, ) of list of float | None
```

Comment:
oh wait, now I understand why you don't use numpy array. It's because each trial has different number of spikes? I think it would still be good to use list of list instead of tuple of list. Tuples are immutable. You'd use them when you don't expect to update the array

Reply:
Thinking a bit more about it, do we even need the "list of list". Why not just a flat list? Do we ever need to know which spikes belong to which trials?

Reply:
> oh wait, now I understand why you don't use numpy array. It's because each trial has different number of spikes? I think it would still be good to use list of list instead of tuple of list. Tuples are immutable. You'd use them when you don't expect to update the array

I like that spike data is stored in tuples because it makes it clear that spike data should only be modified by the `Network` class and should thus remain immutable to the user. The trade-off is that this makes our code a little less sleek. If you prefer otherwise, however, I'm happy to change it to a list of lists.

>Thinking a bit more about it, do we even need the "list of list". Why not just a flat list? Do we ever need to know which spikes belong to which trials?

HNN-gui currently plots spiking activity according to trial so that the user can visualize both aggregate and single-trial spikes. This is reflected in the current HNN I/O convention to write spikes to files on a single-trial basis. Furthermore, if we were ever to attempt running stats on different spiking states that emerge across trials, it would be really important to distinguish between trials.

Reply:
> I like that spike data is stored in tuples because it makes it clear that spike data should only be modified by the Network class and should thus remain immutable to the user.

I see, the way you would do this then is through a private attribute e.g., `_times` and then use a "property" decorator on a method called "times", so the user can only access it but not modify it (since they are not supposed to touch private attributes that start with an underscore). See [here](https://github.com/mne-tools/mne-python/blob/master/mne/epochs.py#L1452-L1455) for an example. I feel mixing tuples and lists might confuse someone reading the code in the future.

Reply:
> HNN-gui currently plots spiking activity according to trial so that the user can visualize both aggregate and single-trial spikes. This is reflected in the current HNN I/O convention to write spikes to files on a single-trial basis. Furthermore, if we were ever to attempt running stats on different spiking states that emerge across trials, it would be really important to distinguish between trials.

okay fair enough!

Reply:
Would we then need to add a `set_times()` function so that `Network` can modify the `Spikes::times` attribute?

Reply:
umm ... I don't think you need it. You are allowed to modify private attributes in the hnn-core code. But the user is not supposed to touch it. A `set_times` might be useful if it's non-trivial to change the `spikes.times` -- e.g., if it's linked to other attributes so to change it safely to add some logic.